### Standardization of SUPERCIAS's data.

The objective of this notebook is to automate the preprocessing and generation of the financial statements of all the companies listed in the Superintendencia de Compañías, Valores y Seguros. The information is quite complete, but the aggregated accounts tend to be imperfectly registered, which is why it's necessary to fill them automatically using the values of the accounts that compose them.

This task can be done using the IFRS manual developed by SUPERCIAS; which details with accuracy which accounts compose the faulty aggregated variables before mentioned. The general idea is to sum up those smaller accounts to get a better version of the faulty variables.

In [1]:
#Import Libraries
import pandas as pd
import os
from zipfile import ZipFile
import tabula

Before working on the standardization it is crucial to find out how does the aggregation method of the Ecuadorian Superintendency of Companies, Stocks, and Insurance works. To do this we decided to grab the largest company, in terms of revenue (i.e. Corporación Favorita C.A.). The reasoning behind this is that such a large company ought to have the least empty values possible for the disaggregated accounts; therefore making it easier to make a complete standardization of the methods needed to generate the aggregated accounts.

We will read the last balance sheet and income statement submitted by this company, then we will use the IFRS instructive developed by the Superintendency to design the aggregation methodology. The idea is to create a function that can take in the value of a specific account, and sum up all the disaggregated "children" of that account.

In [2]:
#Read the submitted statements from the pdf.
fav = tabula.read_pdf("Guide_Docs/favorita_2022.pdf",pages = "all")
for i in range(len(fav)):
    if i == 0:
        num = 7
    else:
        num = 0
    cols = fav[i].iloc[num]
    fav[i].columns = cols
    fav[i] = fav[i].iloc[num+1:].reset_index(drop = True)
fav = fav[:-1]
fav = pd.concat(fav)
#Set dtypes for each column
fav["CUENTA"] = fav["CUENTA"].apply(str)
fav["CÓDIGO"] = fav["CÓDIGO"].apply(str)
fav["VALOR (En USD$)"] = fav["VALOR (En USD$)"].astype(float)
#Visualize data as units
pd.options.display.float_format = '{:.2f}'.format
fav


CUENTA   CÓDIGO  VALOR (En USD$)
0                                     ACTIVO        1    2480403867.05
1                           ACTIVO CORRIENTE      101     701635273.03
2        EFECTIVO Y EQUIVALENTES DE EFECTIVO    10101       7283311.87
3                                       CAJA  1010101       1342366.55
4         INSTITUCIONES FINANCIERAS PÚBLICAS  1010102             0.00
..                                       ...      ...              ...
48  SUPERÁVIT POR REVALUACIÓN DE INVERSIONES    30607             0.00
49                  RESULTADOS DEL EJERCICIO      307     152679114.47
50                 GANANCIA NETA DEL PERIODO    30701     152679114.47
51              (-) PÉRDIDA NETA DEL PERIODO    30702             0.00
52                PARTICIPACIÓN CONTROLADORA       31             0.00

[376 rows x 3 columns]

In [3]:
set(fav["CÓDIGO"].str.len())

{1, 2, 3, 5, 7, 9, 11}

In [4]:
#Display parent
print("Displaying Total Assets account:")
display(fav[fav["CÓDIGO"] == "1"])
#Display children
print("What children is the Total Assets account composed of?")
display(fav[(fav["CÓDIGO"].str.startswith("1")) & (fav["CÓDIGO"].str.len() == 3)])
s = fav[(fav["CÓDIGO"].str.startswith("1")) & (fav["CÓDIGO"].str.len() == 3)]["VALOR (En USD$)"].sum()
print(f"Total sum of children = {s}")
#Test if parent matches the sum of children
print(f"""Is sum of children equal to parent?: {fav[fav["CÓDIGO"] == "1"]["VALOR (En USD$)"][0] == s}""")

Displaying Total Assets account:


CUENTA CÓDIGO  VALOR (En USD$)
0  ACTIVO      1    2480403867.05

What children is the Total Assets account composed of?


CUENTA CÓDIGO  VALOR (En USD$)
1       ACTIVO CORRIENTE    101     701635273.03
3  ACTIVOS NO CORRIENTES    102    1778768594.02

Total sum of children = 2480403867.05
Is sum of children equal to parent?: True


According to the syntax of the account numeration, as well as the IFRS instructive, the accounts follow a specific logic. All accounts have an odd number of digits. Their children contain the same characters as their parents, plus two other characters more. For example, the Total Assets account (1) is composed of the Current Assets (101) and the Non Current Assets (102) accounts. And the same goes for the accounts that compose the current assets and non current assets accounts which are, respectively, (101XX) & (102XX).

Now while this logic works well with some accounts it might fail for those accounts that take away from the total, but are registered with positive values. It is necessary to identify these accounts, preferably in an automated fashion. Then, it will be useful to apply absolute value to each of their values and turn them all negative again; just in case there was an error uploading the numbers.

In [5]:
display(fav[fav["CUENTA"].str.contains("\(-\)")])
#Negative value function
def negate(df):
    p = -abs(df[df["CUENTA"].str.contains(r"\(-\)")]["VALOR (En USD$)"])
    df.loc[df["CUENTA"].str.contains(r"\(-\)"),"VALOR (En USD$)"] = p
negate(fav)
fav[fav["CUENTA"].str.contains("\(-\)")]

CUENTA     CÓDIGO  \
45  (-) PROVISIÓN POR VALOR NETO DE REALIZACIÓN Y ...    1010313   
16  (-) DEPRECIACIÓN ACUMULADA PROPIEDADES, PLANTA...    1020112   
17  (-) DETERIORO  ACUMULADO DE PROPIEDADES, PLANT...    1020113   
20  (-) AMORTIZACION ACUMULADA DE ACTIVOS DE EXPLO...  102011402   
21  (-) DETERIORO  ACUMULADO DE ACTIVOS DE EXPLORA...  102011403   
29  (-) DEPRECIACION ACUMULADA DE PROPIEDADES DE I...    1020203   
30  (-) DETERIORO ACUMULADO DE PROPIEDADES DE INVE...    1020204   
36   (-) DEPRECIACION ACUMULADA DE ACTIVOS BIOLÓGICOS    1020305   
37      (-) DETERIORO ACUMULADO DE ACTIVOS BIOLOGÍCOS    1020306   
43   (-) AMORTIZACIÓN ACUMULADA DE ACTIVOS INTANGIBLE    1020405   
44       (-) DETERIORO ACUMULADO DE ACTIVO INTANGIBLE    1020406   
49  (-) PROVISION POR DETERIORO DE ACTIVOS FINANCI...    1020602   
51  (-) PROVISIÓN POR DETERIORO DE ACTIVOS FINANCI...    1020604   
53  (-)PROVISIÓN POR DETERIORO DE ACTIVOS FINANCIE...    1020606   
0   (-) DEPRECIACIÓN ACUMULADA DE ACTIVOS PROVENIE...    1020701   
1   (-) DETERIORO ACUMULADO DE ACTIVOS PROVENIENTE...    1020702   
12             (-) PROVISIÓN VALUACIÓN DE INVERSIONES    1020810   
25  (-) CAPITAL SUSCRITO NO PAGADO, ACCIONES EN TE...      30102   
43                            (-) PÉRDIDAS ACUMULADAS      30602   
51                       (-) PÉRDIDA NETA DEL PERIODO      30702   

    VALOR (En USD$)  
45             0.00  
16    -476943778.97  
17             0.00  
20             0.00  
21             0.00  
29     -35723314.36  
30             0.00  
36             0.00  
37             0.00  
43     -36427660.12  
44             0.00  
49             0.00  
51             0.00  
53             0.00  
0      -38469261.93  
1              0.00  
12             0.00  
25             0.00  
43      -9793179.94  
51             0.00

CUENTA     CÓDIGO  \
45  (-) PROVISIÓN POR VALOR NETO DE REALIZACIÓN Y ...    1010313   
16  (-) DEPRECIACIÓN ACUMULADA PROPIEDADES, PLANTA...    1020112   
17  (-) DETERIORO  ACUMULADO DE PROPIEDADES, PLANT...    1020113   
20  (-) AMORTIZACION ACUMULADA DE ACTIVOS DE EXPLO...  102011402   
21  (-) DETERIORO  ACUMULADO DE ACTIVOS DE EXPLORA...  102011403   
29  (-) DEPRECIACION ACUMULADA DE PROPIEDADES DE I...    1020203   
30  (-) DETERIORO ACUMULADO DE PROPIEDADES DE INVE...    1020204   
36   (-) DEPRECIACION ACUMULADA DE ACTIVOS BIOLÓGICOS    1020305   
37      (-) DETERIORO ACUMULADO DE ACTIVOS BIOLOGÍCOS    1020306   
43   (-) AMORTIZACIÓN ACUMULADA DE ACTIVOS INTANGIBLE    1020405   
44       (-) DETERIORO ACUMULADO DE ACTIVO INTANGIBLE    1020406   
49  (-) PROVISION POR DETERIORO DE ACTIVOS FINANCI...    1020602   
51  (-) PROVISIÓN POR DETERIORO DE ACTIVOS FINANCI...    1020604   
53  (-)PROVISIÓN POR DETERIORO DE ACTIVOS FINANCIE...    1020606   
0   (-) DEPRECIACIÓN ACUMULADA DE ACTIVOS PROVENIE...    1020701   
1   (-) DETERIORO ACUMULADO DE ACTIVOS PROVENIENTE...    1020702   
12             (-) PROVISIÓN VALUACIÓN DE INVERSIONES    1020810   
25  (-) CAPITAL SUSCRITO NO PAGADO, ACCIONES EN TE...      30102   
43                            (-) PÉRDIDAS ACUMULADAS      30602   
51                       (-) PÉRDIDA NETA DEL PERIODO      30702   

    VALOR (En USD$)  
45            -0.00  
16    -476943778.97  
17            -0.00  
20            -0.00  
21            -0.00  
29     -35723314.36  
30            -0.00  
36            -0.00  
37            -0.00  
43     -36427660.12  
44            -0.00  
49            -0.00  
51            -0.00  
53            -0.00  
0      -38469261.93  
1             -0.00  
12            -0.00  
25            -0.00  
43      -9793179.94  
51            -0.00

#### Designing the aggregation function

In [18]:
#Function that checks if the sum of immediate children is equal to the value of the aggregate account
def checkSum(df,acct):
    df1 = df.copy(deep = True)
    acct = str(acct)
    #Children dframe
    a = df[(df["CÓDIGO"].str.startswith(acct)) & (df["CÓDIGO"].str.len() == len(acct)+2)]
    #Keep track of the number of children available for the account
    print(f"Size of children dframe: {len(a)}")

    o_value = round(df[(df["CÓDIGO"].str.startswith(acct)) & (df["CÓDIGO"].str.len() == len(acct))]["VALOR (En USD$)"].values[0],2)
    n_value = round(a["VALOR (En USD$)"].sum(),2)
    print(f"nvalue {n_value}")
    print("Digits in children")
    display(len(a["CÓDIGO"].tolist()[0]))
    print("Children df:")
    display(a)
    print(f"""Sum of children = {n_value}""")
    print(f"""Actual value of Account: {o_value}""")
    if o_value == n_value:
        print("Sum of children = Aggregated Account")
    else:
        print("Sum of children != Aggregated Account")
    df1.loc[df1["CÓDIGO"] == acct, "VALOR (En USD$)"] = n_value
    stuff = df1[(df1["CÓDIGO"].str.startswith(acct)) & (df1["CÓDIGO"].str.len() == len(acct))]["VALOR (En USD$)"].values[0]
    display(round(stuff,2))
checkSum(fav,10201)

Size of children dframe: 14
nvalue 842886295.83
Digits in children


7

Children df:


CUENTA   CÓDIGO  \
5                                            TERRENOS  1020101   
6                                           EDIFICIOS  1020102   
7                             CONSTRUCCIONES EN CURSO  1020103   
8                                       INSTALACIONES  1020104   
9                                   MUEBLES Y ENSERES  1020105   
10                                MAQUINARIA Y EQUIPO  1020106   
11            NAVES, AEREONAVES, BARCAZAS Y SIMILARES  1020107   
12                              EQUIPO DE COMPUTACIÓN  1020108   
13  VEHÍCULOS, EQUIPOS DE TRASPORTE Y EQUIPO CAMIN...  1020109   
14                 OTROS PROPIEDADES, PLANTA Y EQUIPO  1020110   
15                           REPUESTOS Y HERRAMIENTAS  1020111   
16  (-) DEPRECIACIÓN ACUMULADA PROPIEDADES, PLANTA...  1020112   
17  (-) DETERIORO  ACUMULADO DE PROPIEDADES, PLANT...  1020113   
18               ACTIVOS DE EXPLORACION Y EXPLOTACION  1020114   

    VALOR (En USD$)  
5      198331581.21  
6      364717482.41  
7       35391192.48  
8              0.00  
9       25416497.26  
10     313521038.97  
11             0.00  
12      61823261.94  
13      19212204.50  
14     301416816.03  
15             0.00  
16    -476943778.97  
17            -0.00  
18             0.00

Sum of children = 842886295.83
Actual value of Account: 842886295.83
Sum of children = Aggregated Account


842886295.83

In [17]:
#Function that recursively sums up the childrens of accounts to replace them with the value of the aggregated account

"""It identifies children for an account, and then it identifies the children of the children, recursively, 
until there are no more children, at which point it sums up all of the children to replace this sum with the value of 
the aggregate account"""

def sumup(df,acct):
    acct = str(acct)
    #Children dframe
    children = df[(df["CÓDIGO"].str.startswith(acct)) & (df["CÓDIGO"].str.len() == len(acct)+2)]
    #Check for presence of children
    if children.empty:
        return df[df["CÓDIGO"] == acct]["VALOR (En USD$)"].values[0]
    else:
        sum_children = 0
        for index,child in children.iterrows():
            sum_children += sumup(df,child["CÓDIGO"])

        #If it's the equity account, sum account 31, which was registered in pair numbers of digits for an unknown reason
        if acct == "3":
            sum_children += df[df["CÓDIGO"] == "31"]["VALOR (En USD$)"].values[0]
        #Update the parent's value with the sum of children
        df.loc[df["CÓDIGO"] == acct,"VALOR (En USD$)"] = sum_children
        return round(sum_children,2)
sumup(fav,10201)

842886295.83

With this recursive implementation we ensure that every parent from the balance sheet is calculated based on the sum of their children. This can be used on the massive data from the balances to achieve a more complete information for those companies that have values of zero for their aggregate accounts, but non-zero values for the children accounts.

In [14]:
dic = pd.read_csv("Financials/catalogo_2022_1.txt",sep = "\t",encoding = "latin-1",header = None,)
dic

ValueError: header must be integer or list of integers

In [11]:
#Decompress 2022's financial statements
with ZipFile("Financials/estadosFinancieros_2022.zip","r") as zip_ref:
    zip_ref.extractall("Financials")
#Read Financial Statements ¿
# df = pd.read_csv("Financials/balances_2022_1.txt",sep = "\t",encoding = "latin-1")
# dic = pd.read_csv("Financials/catalogo_2022_1.txt",sep = "\t",encoding = "latin-1")
# display(df.head(5))
# display(dic.head(5))
#Remove files once read by pandas
# for i in [i for i in os.listdir("Financials") if "balances" in i or "catalogo" in i]:
#     os.remove(f"Financials/{i}")

In [15]:
df

AÑO  EXPEDIENTE            RUC  \
0      2022           1  1790013731001   
1      2022           2  1790004724001   
2      2022           3  1790008959001   
3      2022          11  1790044149001   
4      2022          22  1790023516001   
...     ...         ...            ...   
99043  2022      749940  2191774183001   
99044  2022      750412  2191774209001   
99045  2022      750435  1391934838001   
99046  2022      750568  1793207692001   
99047  2022      751125  1091797737001   

                                                  NOMBRE RAMA_ACTIVIDAD  \
0               ACEITES TROPICALES SOCIEDAD ANONIMA ATSA              A   
1                          ACERIA DEL ECUADOR CA ADELCA.              C   
2                       ACERO COMERCIAL ECUATORIANO S.A.              G   
3         AEROVIAS DEL CONTINENTE AMERICANO S.A. AVIANCA              H   
4                AGENCIAS Y REPRESENTACIONES CORDOVEZ SA              G   
...                                                  ...            ...   
99043            COMPAÑIA RENTACAR TRANSANTACECILIA S.A.              N   
99044  COMPAÑIA DE TRANSPORTE PESADO PAUTA & EDA TRAN...              H   
99045                                SYSNET-ECUADOR S.A.              J   
99046  ECUADOR LIFETIME EXPERIENCE LIFETIMECUADOR CIA...              N   
99047                         RUNATRIP ECUADOR CIA.LTDA.              N   

                                        DESCRIPCION_RAMA      CIIU  \
0         AGRICULTURA, GANADERÍA,  SILVICULTURA Y PESCA.  A0126.01   
1                             INDUSTRIAS MANUFACTURERAS.  C2410.25   
2      COMERCIO AL POR MAYOR Y AL POR MENOR REPARACIÓ...  G4610.03   
3                           TRANSPORTE Y ALMACENAMIENTO.  H5110.01   
4      COMERCIO AL POR MAYOR Y AL POR MENOR REPARACIÓ...  G4630.95   
...                                                  ...       ...   
99043  ACTIVIDADES DE SERVICIOS ADMINISTRATIVOS Y DE ...  N7710.00   
99044                       TRANSPORTE Y ALMACENAMIENTO.  H4923.01   
99045                        INFORMACIÓN Y COMUNICACIÓN.  J6110.01   
99046  ACTIVIDADES DE SERVICIOS ADMINISTRATIVOS Y DE ...  N7912.00   
99047  ACTIVIDADES DE SERVICIOS ADMINISTRATIVOS Y DE ...  N7912.00   

           CUENTA_1    CUENTA_101 CUENTA_10101  ... CUENTA_80004 CUENTA_80005  \
0        1360051,92      71630,18     54744,04  ...            0  -1244813,70   
1      459045627,92  241767990,46  41809089,24  ...            0            0   
2       11224470,90   10893439,61   1238466,08  ...            0            0   
3       20915428,96    2181954,71   2050454,71  ...            0            0   
4       27799662,03   27131380,13   1212276,74  ...            0      1402,76   
...             ...           ...          ...  ...          ...          ...   
99043       1200,00       1200,00      1200,00  ...            0            0   
99044       1000,00       1000,00            0  ...            0            0   
99045       1000,00       1000,00      1000,00  ...            0            0   
99046        400,00        400,00       400,00  ...            0            0   
99047        400,00        400,00       400,00  ...            0            0   

      CUENTA_80006 CUENTA_80007 CUENTA_80008 CUENTA_80009   CUENTA_801  \
0                0            0            0            0  -1288545,54   
1                0            0            0            0            0   
2                0            0            0            0            0   
3                0            0            0            0            0   
4                0            0            0            0   1010213,23   
...            ...          ...          ...          ...          ...   
99043            0            0            0            0            0   
99044            0            0            0            0            0   
99045            0            0            0            0            0   
99046            0            0            0    

In [153]:
abs(p["VALOR (En USD$)"])

45           0.00
16   476943778.97
17           0.00
20           0.00
21           0.00
29    35723314.36
30           0.00
36           0.00
37           0.00
43    36427660.12
44           0.00
49           0.00
51           0.00
53           0.00
0     38469261.93
1            0.00
12           0.00
25           0.00
43     9793179.94
51           0.00
Name: VALOR (En USD$), dtype: float64